Support initializing accelerators of specific kinds

tungld opened the pull request:
Currently, `initAccelerators` function initializes all built accelerators, which is redundant when `--maccel` is specified for a subset of accelerators.

This patch edits  `initAccelerators` to support initializing only accelerators of specific kinds.

This patch also adds a new OptionCategory that is `OnnxMlirCommonOptions` for the options shared between `onnx-mlir` and `onnx-mlir-opt`. In `onnx-mlir-opt`, we enable only options in `OnnxMlirCommonOptions` so that the other unrelated options will not show up, which makes the output of `onnx-mlir-opt --help` clean.

Resolves #1406

Signed-off-by: Tung D. Le <tung@jp.ibm.com>

AlexandreEichenberger commented:
Is it ok to let #1395 go first?

tungld commented:
> Is it ok to let https://github.com/onnx/onnx-mlir/pull/1395 go first?

Yes, it's totally ok.

AlexandreEichenberger commented on src/Conversion/KrnlToLLVM/ConvertKrnlToLLVM.cpp:
```diff
@@ -520,12 +520,13 @@ void ConvertKrnlToLLVMPass::runOnOperation() {
   target.addLegalOp<UnrealizedConversionCastOp>();
 
+  // Get active accelerators if any.
+  SmallVector<onnx_mlir::accel::Accelerator *, 2> activeAccels;
```
Out of curiosity, would it be sufficient to return a reference to the class list with a const?
`const SmallVectorImpl<onnx_mlir::accel::Accelerator>& getActiveAccelerator()`

tungld replied:
Removed this function since it becomes redundant now.

tungld commented on src/Accelerators/Accelerator.hpp:
```diff
@@ -28,7 +28,10 @@
 #define DECLARE_ACCEL_INIT_FUNCTION(name) extern Accelerator *create##name();
-#define INVOKE_ACCEL_INIT_FUNCTION(name) create##name()->setName(#name);
+#define INVOKE_ACCEL_INIT_FUNCTION(name, kinds)                                \
+  if (kinds.empty() ||                                                         \
```
> I don't understand why we support empty kinds (which is maccel) differently?

`onnx-mlir-opt` will initialize all accelerators so that accelerators' passes will be available. If you prefer, we can allow `-maccel` in `onnx-mlir-opt` so that only passes for the given accelerators are available, and empy kind means there is no accelerator.

> I would rather allow building for NNPA on a non-Z machine and do a lit test on a x86 machine by expressing the -maccel=NNPA in the onnx-mlir or onnx-mlir-opt.

This is another thing. It depends on whether we build libzdnn or not. If not, we can do the lit test on a x86 machine. A PR that allows turning on/off libzdnn build can solve the problem.

gongsu832 replied:
> There is something funny about the lit-tests, and maybe that is the reason of the confusion? In the makefiles, we disable building for NNPA if we are not on an Z machine... and maybe that is the source of the problem? I would rather allow building for NNPA on a non-Z machine and do a lit test on a x86 machine by expressing the `-maccel=NNPA` in the `onnx-mlir` or `onnx-mlir-opt`.

There is no reason to build NNPA on non-Z machines since it will never run there, and that's why PR #1384 was created to allow NNPA lit tests on non-Z machines even if NNPA cannot be built.

AlexandreEichenberger replied:
>  If you prefer, we can allow -maccel in onnx-mlir-opt so that only passes for the given accelerators are available, and empy kind means there is no accelerator.

yes, that would be my preference. That way things looks more like `onnx-mlir`.

tungld commented on src/Compiler/CompilerUtils.cpp:
```diff
@@ -780,23 +779,25 @@ int compileModule(mlir::OwningOpRef<ModuleOp> &module,
-      if (!accel->isActive())
-        continue;
-      hasActiveAccel = true;
+  // TODO(tung): Revise adding passes. The current mechanism does not work if
```
I added an assertion inside `getAccelerators()` so that the code with loops is unchanged. Thanks!

tungld commented:
@AlexandreEichenberger I addressed your comments. Please take a look.
While adding `--maccel` to `onnx-mlir-opt`, I reorganized options a bit so that only necessary options are available for `onnx-mlir-opt`, e.g. OptimizationLevel and maccel options.

tungld commented on src/Tools/onnx-mlir-opt/onnx-mlir-opt.cpp:
```diff
@@ -81,7 +81,27 @@ void scanAndSetOptLevel(int argc, char **argv) {
 }
 
+void scanAndSetMAccel(int argc, char **argv) {
+  // Scan accelerators and add them to the maccel option.
+  for (int i = argc - 1; i > 0; --i) {
+    std::string currStr(argv[i]);
+    if (currStr.find("--maccel=") != 0)
+      continue;
+    std::string accelKind(&argv[i][9]); // Get the string starting 9 chars down.
+    setTargetAccel(accelKind);
+  }
+}
+
```
There are two kinds of options: one includes `--maccel`, `-O{0,1,2,3}` and the other includes pass options such as `---shape-inference`, `--convert-onnx-to-zhigh`, etc. 

There is a dependency here that is pass options requires passes being registered, and to register passes we need `--maccel`, `-O{0,1,2,3}`, that is why we have to parse the command line manually.

Below is the prototype in the main function of `onnx-mlir-opt`, which may help us understand visually what I mentioned. 
```
  // Cannot call llvm::cl::ParseCommandLineOptions here because pass options have not been not available yet and ParseCommandLine would fail in parsing pass options.

  // Scan Opt Level manually now as it is needed for initializing the OM Passes.
  scanAndSetOptLevel(argc, argv);
  // Scan maccel manually now as it is needed for initializing the OM Passes.
  scanAndSetMAccel(argc, argv);

  // Register passes
  onnx_mlir::initOMPasses(OptimizationLevel);

  // Now passes are intialized, parse commandline. 
  mlir::PassPipelineCLParser passPipeline("", "Compiler passes to run");  // these will pull pass options like `--shape-inference`, `--convert-onnx-to-zhigh`, etc.
  llvm::cl::ParseCommandLineOptions(argc, argv,
      "ONNX-MLIR modular optimizer driver\n", nullptr, "ONNX_MLIR_OPT_FLAGS");
```

etiotto replied:
> There is a dependency here that is pass options requires passes being registered, and to register passes we need --maccel, -O{0,1,2,3}, that is why we have to parse the command line manually.

Now I understand the reason for that, thanks!  Suggest to put a comment in the code to state that so we remember :-)

tungld replied:
We've already had it in the code :)
```
  // Scan Opt Level manually now as it is needed for initializing the OM Passes.
  scanAndSetOptLevel(argc, argv);
  // Scan maccel manually now as it is needed for initializing the OM Passes.
  scanAndSetMAccel(argc, argv);
```

etiotto reviewed (approved):
LGTM once @AlexandreEichenberger comments are addressed

tungld commented:
> LGTM once @AlexandreEichenberger comments are addressed

@AlexandreEichenberger I addressed your comments, please give this another review. Thanks!

tungld commented on src/Accelerators/Accelerator.hpp:
```diff
@@ -144,5 +149,7 @@ std::ostream &operator<<(std::ostream &out, const Accelerator::Kind kind);
 llvm::raw_ostream &operator<<(
     llvm::raw_ostream &out, const Accelerator::Kind kind);
 
+extern void initAccelerators(llvm::ArrayRef<Accelerator::Kind> kinds = {});
```
Removed this. Thanks!

tungld commented on src/Accelerators/NNPA/NNPAAccelerator.cpp:
```diff
@@ -55,9 +55,7 @@ NNPAAccelerator::NNPAAccelerator() : Accelerator(Accelerator::Kind::NNPA) {
-  if (instance || llvm::any_of(maccel, [](Accelerator::Kind kind) {
-        return kind == Accelerator::Kind::NNPA;
-      })) {
+  if (instance) {
```
We have no concept of active accelerator as we discussed in the last meeting.

AlexandreEichenberger reviewed (approved):
LGTM, looks much much better, tx
